File: flitetrakr_sync/src/integrationTest/java/com/github/albertosh/flitetrakr/sync/usecases/connectionsbelowprice/ConnectionsBelowPriceUseCaseTest.java
package com.github.albertosh.flitetrakr.sync.usecases.connectionsbelowprice;

import com.github.albertosh.flitetrakr.model.Connection;
import com.github.albertosh.flitetrakr.model.MultipleConnections;
import com.github.albertosh.flitetrakr.sync.services.ConnectionService;
import com.github.albertosh.flitetrakr.sync.services.IConnectionService;
import com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection.CheapestConnectionUseCase;
import com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection.CheapestConnectionUseCaseError;
import com.github.albertosh.flitetrakr.sync.usecases.cheapestconnection.ICheapestConnectionUseCase;

import org.junit.Before;
import org.junit.Test;

import java.util.List;

import static org.hamcrest.CoreMatchers.not;
import static org.hamcrest.CoreMatchers.nullValue;
import static org.hamcrest.Matchers.contains;
import static org.hamcrest.collection.IsCollectionWithSize.hasSize;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;

public class ConnectionsBelowPriceUseCaseTest {

    private IConnectionService service;
    private ICheapestConnectionUseCase cheapestConnectionUseCase;
    private ConnectionsBelowPriceUseCase connectionsBelowPriceUseCase;

    @Before
    public void setUp() throws Exception {
        service = new ConnectionService();
        cheapestConnectionUseCase = new CheapestConnectionUseCase(service);
        connectionsBelowPriceUseCase = new ConnectionsBelowPriceUseCase(cheapestConnectionUseCase, service);

        service.addConnection(new Connection("NUE-FRA-43"));
        service.addConnection(new Connection("NUE-AMS-67"));
        service.addConnection(new Connection("FRA-AMS-17"));
        service.addConnection(new Connection("FRA-LHR-27"));
        service.addConnection(new Connection("LHR-NUE-23"));
    }

    @Test
    public void existingConnections() throws Exception, CheapestConnectionUseCaseError, ConnectionsBelowPriceUseCaseError {
        final String FROM = "NUE";
        final String TO = "LHR";
        final Integer PRICE = 170;

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();
        List<MultipleConnections> output = connectionsBelowPriceUseCase.execute(input);

        assertThat(output, hasSize(2));

        assertThat(output.get(0).getCities(), contains("NUE", "FRA", "LHR"));
        assertThat(output.get(0).getPrice(), is(70));

        assertThat(output.get(1).getCities(), contains("NUE", "FRA", "LHR", "NUE", "FRA", "LHR"));
        assertThat(output.get(1).getPrice(), is(163));
    }

    @Test
    public void whenTheCheapestConnectionIsOverMinPriceTheAlgorithmHalts() throws Exception, CheapestConnectionUseCaseError, ConnectionsBelowPriceUseCaseError {
        final String FROM = "NUE";
        final String TO = "LHR";
        final Integer PRICE = 60;

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();
        List<MultipleConnections> output = connectionsBelowPriceUseCase.execute(input);

        assertThat(output, hasSize(0));
    }

    @Test
    public void aNonExistingConnectionsThrowsAnError() throws Exception, CheapestConnectionUseCaseError {
        final String FROM = "NUE";
        final String TO = "DBX";
        final Integer PRICE = 1000;

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();

        ConnectionsBelowPriceUseCaseError error = null;
        try {
            connectionsBelowPriceUseCase.execute(input);
        } catch (ConnectionsBelowPriceUseCaseError connectionsBelowPriceUseCaseError) {
            error = connectionsBelowPriceUseCaseError;
        }

        assertThat(error, is(not(nullValue())));
        assertThat(error.getErrorType(), is(ConnectionsBelowPriceUseCaseError.ErrorType.CONNECTION_NOT_FOUND));
    }

    @Test
    public void sameFromAndTo() throws Exception, CheapestConnectionUseCaseError, ConnectionsBelowPriceUseCaseError {
        final String FROM = "LHR";
        final String TO = "LHR";
        final Integer PRICE = 100;

        ConnectionsBelowPriceUseCaseInput input = new ConnectionsBelowPriceUseCaseInput.Builder()
                .from(FROM)
                .to(TO)
                .price(PRICE)
                .build();
        List<MultipleConnections> output = connectionsBelowPriceUseCase.execute(input);

        assertThat(output, hasSize(2));

        assertThat(output.get(0).getCities(), contains("LHR"));
        assertThat(output.get(0).getPrice(), is(0));

        assertThat(output.get(1).getCities(), contains("LHR", "NUE", "FRA", "LHR"));
        assertThat(output.get(1).getPrice(), is(93));
    }
}

File: flitetrakr_reactive_kotlin/src/integrationTest/kotlin/com/github/albertosh/flitetrakr/usecases/ConnectionsBelowPriceUseCaseTest.kt
package com.github.albertosh.flitetrakr.usecases

import com.github.albertosh.flitetrakr.model.MultipleConnections
import com.github.albertosh.flitetrakr.services.IConnectionService
import io.kotlintest.specs.ShouldSpec

class ConnectionsBelowPriceUseCaseTest : ShouldSpec() {

    init {
        "ConnectionsBelowPriceUseCase" {
            should("throw an error if a connection doesn't exist") {
                val service: IConnectionService = buildService()
                val cheapestConnectionUseCase = CheapestConnectionUseCase(service)
                val useCase = ConnectionsBelowPriceUseCase(service, cheapestConnectionUseCase)

                useCase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "NUE",
                                to = "DBX",
                                price = Int.MAX_VALUE
                        ))
                        .test()
                        .assertError(ConnectionsBelowPriceUseCaseError.connectionNotFound)

            }

            should("look for connections below some price if at least a connection below that price exists") {
                val service: IConnectionService = buildService()
                val cheapestConnection = CheapestConnectionUseCase(service)
                val usecase = ConnectionsBelowPriceUseCase(service, cheapestConnection)

                usecase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "NUE",
                                to = "LHR",
                                price = 170
                        ))
                        .test()
                        .assertValues(
                                ConnectionsBelowPriceUseCaseOutput(
                                        connections = listOf(
                                                MultipleConnections(
                                                        listOf("NUE", "FRA", "LHR"),
                                                        70)
                                        )),
                                ConnectionsBelowPriceUseCaseOutput(
                                        connections = listOf(
                                                MultipleConnections(
                                                        listOf("NUE", "FRA", "LHR"),
                                                        70),
                                                MultipleConnections(
                                                        listOf("NUE", "FRA", "LHR", "NUE", "FRA", "LHR"),
                                                        163)
                                        )
                                ))
                        .assertComplete()

            }

            should("halt when the cheapest connection is over maximum price") {
                val service: IConnectionService = buildService()
                val cheapestConnectionUseCase = CheapestConnectionUseCase(service)
                val useCase = ConnectionsBelowPriceUseCase(service, cheapestConnectionUseCase)

                useCase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "NUE",
                                to = "LHR",
                                price = 60))
                        .test()
                        .assertValues(
                                ConnectionsBelowPriceUseCaseOutput(emptyList<MultipleConnections>())
                        )
                        .assertComplete()
            }

            should("return the existing connections ordered below some price even if the destination and origin is the same") {
                val service: IConnectionService = buildService()
                val cheapestConnectionUseCase = CheapestConnectionUseCase(service)
                val useCase = ConnectionsBelowPriceUseCase(service, cheapestConnectionUseCase)

                useCase.execute(
                        ConnectionsBelowPriceUseCaseInput(
                                from = "LHR",
                                to = "LHR",
                                price = 100))
                        .test()
                        .assertValues(
                                ConnectionsBelowPriceUseCaseOutput(
                                        connections = listOf(
                                                MultipleConnections(
                                                        listOf("LHR"),
                                                        0)
                                        )),
                                ConnectionsBelowPriceUseCaseOutput(
                                        connections = listOf(
                                                MultipleConnections(
                                                        listOf("LHR"),
                                                        0),
                                                MultipleConnections(
                                                        listOf("LHR", "NUE", "FRA", "LHR"),
                                                        93
                                                )
                                        )
                                ))
                        .assertComplete()

            }

        }
    }
}
